fix: field replace

blaggacao opened the pull request:
- fix: section fields_dict
- fix: replace_field

Prior to this PR, replace_field did simply not work. Because:
- splice needs an (integer) index not an element
- also section field management must be taken into account for all sections

This PR uses the correct integer for splicing and also manages sections, columns and tabs
according to the original field that is being replaced so that the field remains at the same
location in the layout.

---

works blissfully in tandem with #21322 :
```js
	setup(frm) {
		df = frm.get_docfield("map");
		df["fieldtype"] = "GeolocationVRP";
		frm.layout.replace_field("map", df);
	},
	refresh(frm) {
		fld = frm.fields_dict["map"];
		fld.refresh(); // not sure why necessary, may be rather a bug in ControlGeolocation
	}
```

And a custom controller:
```js
frappe.ui.form.ControlGeolocationVRP = class ControlGeolocationVRP extends (
	frappe.ui.form.ControlGeolocation
) {
	point_to_layer(geoJsonPoint, latlng) {
		// Custom rules for how geojson data is rendered on the map
		if (geoJsonPoint.properties.point_type == "circle") {
			return L.circle(latlng, { radius: geoJsonPoint.properties.radius });
		} else if (geoJsonPoint.properties.point_type == "circlemarker") {
			return L.circleMarker(latlng, {
				radius: geoJsonPoint.properties.radius,
			});
		} else {
			// return L.marker(latlng);
			const homeIcon = L.divIcon({ className: "fa fa-home --blue" });
			return L.marker(latlng, { icon: homeIcon });
		}
	}
	setStyle(geoJsonFeature) {
		if (geoJsonFeature.geometry.type === "LineString") {
			return {
				color: geoJsonFeature.properties.stroke,
				weight: geoJsonFeature.properties["stroke-width"],
			};
		}
		return {};
	}
};
```

blaggacao commented:
something's still missing here.

blaggacao committed:
fix: replace field replacing correct dom element

blaggacao commented:
This was the missing bit: ~~https://github.com/frappe/frappe/pull/21419/commits/7592cbedc36e414acd158d3abc3e84bd5dfb3533~~ https://github.com/frappe/frappe/pull/21419/commits/ecec197bc3a3bf04749344e7be54a4f2a654491a

blaggacao commented:
@shariquerik I just rebased on latest `develop`. So this branch now represents the minimal change set.

barredterra commented on frappe/public/js/frappe/form/section.js:
```diff
@@ -82,6 +82,16 @@ export default class Section {
 	}
 
+	replace_field(fieldname, fieldobj) {
+		if (this.fields_dict[fieldname] && this.fields_dict[fieldname].df) {
```
```suggestion
		if (this.fields_dict[fieldname]?.df) {
```

blaggacao replied:
Ah nice, it's like Nix :smile:

barredterra reviewed (commented):
From the PR description, it's not clear to me, what you are trying to achieve with this PR.

Some nitpicks:

- The PR title should follow the [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/
) style
- The variable names (`fieldobj`, `oldfieldobj`, `olfldobj`) are hard to read and make sense of

blaggacao commented:
@barredterra this is a fix for an issue not covered by a test case. I'd usually have supplied a failing one, but I'm not yet proficient enough, so bear with me :smile:

~~I'll double check on CC commit style. I thought I had checked with `pre-commit`. Double checking...~~

Ah! PR Title! Yeah, saw a reference to some pecularities  somewhere, thanks for reminding me!

Variable names: ack! :eyes:

blaggacao commented:
How could I _unit_-test DOM initialization? I'm not sure how much of a public API the implementation exposes.

While looking at it, it seemed all a bit private api, so not sure how I'd approach a testcase without being sent on an integration test tangent. But maybe there's a way?

blaggacao committed:
fix: replace_field

splice needs an (integer) index not an element

also section field management must be taken into account for all sections

blaggacao commented:
@shariquerik @barredterra So what's the game plan for this PR?

- Rebased on develop
- Attended all reviewer feedback (or so I hope :smile:)

Imo, this is ready to merge.

shariquerik commented on frappe/public/js/frappe/form/layout.js:
```diff
@@ -191,12 +191,14 @@ frappe.ui.form.Layout = class Layout {
-			this.section.fields_list.splice(this.section.fields_dict[fieldname], 1, fieldobj);
-			this.section.fields_dict[fieldname] = fieldobj;
+			this.sections.forEach((section) => section.replace_field(fieldname, fieldobj));
+			prev_fieldobj.tab.replace_field(fieldobj);
```
What if we do not have tabs? this line will break?

shariquerik commented on frappe/public/js/frappe/form/layout.js:
```diff
@@ -191,12 +191,14 @@ frappe.ui.form.Layout = class Layout {
-			this.section.fields_list.splice(this.section.fields_dict[fieldname], 1, fieldobj);
-			this.section.fields_dict[fieldname] = fieldobj;
+			this.sections.forEach((section) => section.replace_field(fieldname, fieldobj));
+			prev_fieldobj.tab.replace_field(fieldobj);
```
```suggestion
			prev_fieldobj.tab?.replace_field(fieldobj);
```

shariquerik replied:
If form has tabs form is breaking

shariquerik committed:
fix: avoid this line if there is no tab break field in the form

shariquerik committed:
Merge branch 'develop' into fix-field-replace

blaggacao commented:
@shariquerik Thanks for the fix! And thanks for motion to merge!

shariquerik commented:
@blaggacao Can you resolve the conflicts in backported PR?
